File: erpnext/accounts/page/daily_sales_report/daily_sales_report.py
from __future__ import unicode_literals
import json
import pdfkit, os, frappe
from frappe.utils import cstr
from frappe import _
from frappe.utils.pdf import get_pdf
from frappe.utils.csvutils import UnicodeWriter
import datetime

@frappe.whitelist()
def single_emp_sales_details(emp=None, from_date=None, to_date=None, mode_of_pay=None):
	employee_query = """	select DATE_FORMAT(s.posting_date,'%d-%m-%Y') as posting_date, i.emp, 
					case when i.income_account like 'Sales -%' 
						then sum(i.amount) else 0.00 END as tot_sales, 
					case when i.income_account like 'Service -%' 
						then sum(i.amount) else 0.00 END as tot_service 
				from 
					`tabSales Invoice Item` i, 
					`tabSales Invoice` s 
				where 
					s.name = i.parent
				and
					s.docstatus = 1 
				and
					(i.income_account like 'Sales -%' or i.income_account like 'Service -%')
			"""
	
	if emp: employee_query += " and i.emp = '%s'"%(emp)
	
	if from_date:
		from_date = datetime.datetime.strptime(from_date, '%d-%m-%Y').strftime('%Y-%m-%d')
		employee_query += " and s.posting_date >= '{0}'".format(from_date)

	if to_date:
		to_date = datetime.datetime.strptime(to_date, '%d-%m-%Y').strftime('%Y-%m-%d')
		employee_query += " and s.posting_date <= '{0}'".format(to_date)

	employee_query += " group by s.posting_date,i.emp,i.income_account"

	sales_details = frappe.db.sql(employee_query,as_dict=True)




	sales_col_tot = service_col_tot = 0.00
	for d in sales_details:
		sales_col_tot += d['tot_sales']
		service_col_tot += d['tot_service']

	for i in sales_details:
		i["tot_sales"] = '%.3f' % i["tot_sales"]
		i["tot_service"] = '%.3f' % i["tot_service"]
		
	sales_col_tot_custom = '%.3f' % sales_col_tot
	total = [{'sales_col_tot': sales_col_tot_custom, "service_col_tot": '%.3f' % service_col_tot}]
	return sales_details, total

@frappe.whitelist()
def get_mode_of_pay_details(from_date=None, to_date=None, mode_of_pay=None, is_sale=None,is_service=None):
	mode_query = """select
				m.mode_of_payment, sum(m.amount) as amount
			from
				`tabSales Invoice` s 
			left join
				`tabMode of Pay` m
			on
				s.name = m.parent
			where
				s.docstatus = 1
		"""

	if mode_of_pay: mode_query += " and m.mode_of_payment = '{0}'".format(mode_of_pay)
	if from_date:
		from_date = datetime.datetime.strptime(from_date, '%d-%m-%Y').strftime('%Y-%m-%d')
		mode_query += " and s.posting_date >= '{0}'".format(from_date)
	
	if to_date:
		to_date = datetime.datetime.strptime(to_date, '%d-%m-%Y').strftime('%Y-%m-%d')
		mode_query += " and s.posting_date <= '{0}'".format(to_date)

	mode_query += " group by m.mode_of_payment"
	mode_of_pay_details = frappe.db.sql(mode_query,as_dict=1)
	tot_amt = 0.00
	for amt in mode_of_pay_details:
		if amt['amount']:
			tot_amt += amt['amount']
			amt['amount'] = '%.3f' % amt['amount']
	total = [{'total': '%.3f' %  tot_amt}]
	return mode_of_pay_details, total

@frappe.whitelist()
def create_csv(emp_data, mode_of_pay,filters):
	w = UnicodeWriter()
	w = add_header(w)
	w = add_data(w, emp_data, mode_of_pay, filters)
	# write out response as a type csv
	frappe.response['result'] = cstr(w.getvalue())
	frappe.response['type'] = 'csv'
	frappe.response['doctype'] = "Daily Sales Report"

def add_header(w):
	w.writerow(["Daily Sales Report"])
	return w

def add_data(w,emp_data, mode_of_pay, filters):
	emp_data = json.loads(emp_data)
	mode_of_pay = json.loads(mode_of_pay)
	filters = json.loads(filters)

	print "is sale", filters[0]['is_sale']
	w.writerow('')
	w.writerow(['', 'Employee',filters[0]['emp'],'', 'From Date',filters[0]['from_date'], '','To Date', filters[0]['to_date'],'', 'Mode of Payment', filters[0]['mode_of_pay']])
	w.writerow('\n')
	w.writerow(['Employee Sales/Service Details'])
	if emp_data[0][0]:
		if filters[0]['is_sale'] == 1 and filters[0]['is_service'] == 0:
			w.writerow(['', 'Date','','Employee', '','Total Sales'])
			for i in emp_data[0]:
				if float(i['tot_sales'])>0:
					row = ['',i['posting_date'], '', i['emp'], '', i['tot_sales']]
					w.writerow(row)
				row = ['','Total','','','', emp_data[1][0]['sales_col_tot'], '']
		elif filters[0]['is_sale'] == 0 and filters[0]['is_service'] == 1:
			w.writerow(['', 'Date','','Employee', '','Total Service'])
			for i in emp_data[0]:
				if float(i['tot_service'])>0:
					row = ['',i['posting_date'], '', i['emp'], '', i['tot_service']]
					w.writerow(row)
				row = ['','Total','','','',  emp_data[1][0]['service_col_tot']]
		else:
			w.writerow(['', 'Date','','Employee', '','Total Sales', '','Total Service'])
			for i in emp_data[0]:
				row = ['',i['posting_date'], '', i['emp'], '', i['tot_sales'], '', i['tot_service']]
				w.writerow(row)
				row = ['','Total','','','', emp_data[1][0]['sales_col_tot'], '', emp_data[1][0]['service_col_tot']]
		w.writerow('\n')
		# row = ['','Total','','','', emp_data[1][0]['sales_col_tot'], '', emp_data[1][0]['service_col_tot']]
		w.writerow(row)
	else:
		w.writerow('')
		w.writerow(['No Data Found'])
	w.writerow('\n')

	w.writerow(['Mode Of Payment Details'])
	if mode_of_pay[0]:
		w.writerow(['', 'Mode Of Payment','','Amount'])
		for i in mode_of_pay[0]:
			if i['mode_of_payment']:
				row = ['',i['mode_of_payment'], '', i['amount']]
				w.writerow(row)
		w.writerow('\n')
		w.writerow(['', 'Total','',mode_of_pay[1][0]['total']])
	else:
		w.writerow('')
		w.writerow(['No Data Found'])

	return w

File: erpnext/accounts/page/daily_sales_report/daily_sales_report.js
frappe.pages['daily-sales-report'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Daily Sales Report',
		single_column: true
	});

	wrapper.daily_sales_report = new daily_sales_report(wrapper)
	frappe.breadcrumbs.add("Accounts");
}

daily_sales_report = Class.extend({
	init: function(wrapper) {
		var me = this;
		this.wrapper_page = wrapper.page;
		this.page = $(wrapper).find('.layout-main-section-wrapper');
		//this.page = $(wrapper).find('.page-content');
		this.wrapper = $(wrapper).find('.page-content');
		this.make_page();
		this.render_payments_details();
		this.set_print_button();
	},

	set_print_button: function() {
		var me = this;
		me.wrapper_page.set_primary_action(__("Export"), function() {
			var employee = me.employee.$input.val()
			var from_date = me.from_date.$input.val()
			var to_date = me.to_date.$input.val()
			var mode_of_pay = me.mode.$input.val()
			var is_sale = me.is_sale.last_value
			var is_service = me.is_service.last_value
			filters = [{'emp': employee, 'from_date': from_date, "to_date": to_date, "mode_of_pay": mode_of_pay, "is_sale": is_sale, "is_service": is_service}]
			window.location.href = repl(frappe.request.url +'?cmd=%(cmd)s&emp_data=%(emp_data)s&mode_of_pay=%(mode_of_pay)s&filters=%(filters)s', {
				cmd: "erpnext.accounts.page.daily_sales_report.daily_sales_report.create_csv",
				emp_data:JSON.stringify(me['emp_data']),
				mode_of_pay: JSON.stringify(me['mode_of_pay']),
				filters: JSON.stringify(filters)
			});
		})
	},

	make_page: function() {
		var me = this;
		html = frappe.render_template("daily_sales_report")
		me.page.html(html)
	},

	render_payments_details: function() {
		var me = this;
		me.single_emp_sales();
		me.get_mode_of_pay_details();
	},

	single_emp_sales: function() {
		var me = this;
		me.from_date = frappe.ui.form.make_control({
			parent: me.page.find(".from_date"),
			df: {
				fieldtype: "Date",
				fieldname: "from_date",
				placeholder: "Date",
			},
			render_input: true
		});
		me.from_date.refresh();

		me.to_date = frappe.ui.form.make_control({
			parent: me.page.find(".to_date"),
			df: {
				fieldtype: "Date",
				fieldname: "to_date",
				placeholder: "Date",
			},
			render_input: true
		});
		me.to_date.refresh();

		me.employee = frappe.ui.form.make_control({
			parent: me.page.find(".employee"),
			df: {
			fieldtype: "Link",
			options: "Employee",
			fieldname: "employee",
			placeholder: "Select Employee"
			},
			render_input: true
		});
		me.employee.refresh();

		me.mode = frappe.ui.form.make_control({
			parent: me.page.find(".mode"),
			df: {
			fieldtype: "Link",
			options: "Mode of Payment",
			fieldname: "employee",
			placeholder: "Mode"
			},
			render_input: true
		});
		me.mode.refresh()

		me.is_sale = frappe.ui.form.make_control({
			parent: me.page.find(".is_sale"),
			df: {
			fieldtype: "Check",
			fieldname: "is_sale",
			},
			// render_input: true
		});
		me.is_sale.value = true,
		me.is_sale.refresh()

		me.is_service = frappe.ui.form.make_control({
			parent: me.page.find(".is_service"),
			df: {
			fieldtype: "Check",
			fieldname: "is_service",
			placeholder: "Mode"
			},
			render_input: true
		});
		me.is_service.value = true,
		me.is_service.refresh()


		me.single_emp_sales_details();
		me.change_from_date();
		me.change_to_date()
		me.change_is_sale()
		me.change_is_service()
		me.change_employee();
	},

	single_emp_sales_details: function() {
		var me = this;
		var employee = me.employee.$input.val()
		var from_date = me.from_date.$input.val()
		var to_date = me.to_date.$input.val()
		var mode_of_pay = me.mode.$input.val()
		var is_sale = me.is_sale.last_value
		var is_service = me.is_service.last_value

		frappe.call({
			method: "erpnext.accounts.page.daily_sales_report.daily_sales_report.single_emp_sales_details",
			args: {"emp": employee, "from_date": from_date, 'to_date': to_date, 'mode_of_pay': mode_of_pay},
			callback: function(r) {
				me.page.find(".single_emp_data").empty();
				if(r.message[0][0]){
					me.emp_data = r.message
					mydata= r.message[0]
					blanklist= []
					if(is_sale==1 && is_service == 0){
						for(i=0;i<mydata.length;i++){
							if(flt(mydata[i]['tot_sales'])!=0){
								blanklist.push(mydata[i])
							}
						}
						me.page.find(".single_emp_data").append(frappe.render_template("single_emp_data", {"data":blanklist, "total": r.message[1], "case1":2}))
					}
					else if(is_sale==0 && is_service == 1){
						for(i=0;i<mydata.length;i++){
							if(flt(mydata[i]['tot_service'])!=0){
								blanklist.push(mydata[i])
							}
						}
						me.page.find(".single_emp_data").append(frappe.render_template("single_emp_data", {"data":blanklist, "total": r.message[1], "case1":3}))

					}
					else if(is_sale==1 && is_service == 1){
						me.page.find(".single_emp_data").append(frappe.render_template("single_emp_data", {"data":mydata, "total": r.message[1], "case1":1}))
					}
					else {
						me.page.find(".single_emp_data").append(frappe.render_template("single_emp_data", {"data":blanklist, "total": r.message[1], "case1":4}))
					}
				}
				else
					me.page.find(".single_emp_data").append("<div class='text-muted text-center' style='font-weight: bold; padding-top: 100px'>No Data Found</div>")
				me.get_mode_of_pay_details();
			}
		})
	},

	get_mode_of_pay_details: function() {
		var me = this;
		var from_date = me.from_date.$input.val()
		var to_date = me.to_date.$input.val()
		var mode_of_pay = me.mode.$input.val()
		var is_sale = me.is_sale.$input.val()
		var is_service = me.is_service.$input.val()


		frappe.call({
			method:"erpnext.accounts.page.daily_sales_report.daily_sales_report.get_mode_of_pay_details",
			args: {"from_date": from_date, 'to_date': to_date, 'mode_of_pay': mode_of_pay, "is_sale": is_sale, "is_service":is_service},
			callback: function(r) {
				me.mode_of_pay = r.message
				me.page.find(".mode_of_pay_data").empty();
				if(r.message[0][0])
					me.page.find(".mode_of_pay_data").append(frappe.render_template("mode_of_payment", {"data": r.message[0], "total": r.message[1]}));
				else
					me.page.find(".mode_of_pay_data").append("<div class='text-muted text-center' style='font-weight: bold; padding-top: 100px'>No Data Found</div>")
				me.mode_of_payment_change();
			}
		})
	},

	change_employee: function() {
		var me = this;
		me.employee.$input.on("change", function(){
			me.single_emp_sales_details();
		})
	},

	change_from_date: function() {
		var me = this;
		me.from_date.$input.on("change", function() {
			from_date = me.from_date.$input.val();
			to_date = me.to_date.$input.val();
			if((from_date && to_date) && from_date > to_date) {
				msgprint("From Date Must be smaller than To Date")
				return false
			}
			else {
				me.single_emp_sales_details();
				me.get_mode_of_pay_details();
			}
		})
	},

	change_to_date: function() {
		var me = this;
		me.to_date.$input.on("change", function() {
			from_date = me.from_date.$input.val();
			to_date = me.to_date.$input.val();

			if((from_date && to_date) && to_date < from_date) {
				msgprint("To Date Must be greater than From Date")
				return false
			}
			else {
				me.single_emp_sales_details();
				me.get_mode_of_pay_details();
			}
		})
	},
	change_is_sale: function() {
		var me = this;
		me.is_sale.$input.on("change", function() {
			from_date = me.from_date.$input.val();
			to_date = me.to_date.$input.val();

			if((from_date && to_date) && to_date < from_date) {
				msgprint("To Date Must be greater than From Date")
				return false
			}
			else {
				me.single_emp_sales_details();
				me.get_mode_of_pay_details();
			}
		})
	},
	change_is_service: function() {
		var me = this;
		me.is_service.$input.on("change", function() {
			from_date = me.from_date.$input.val();
			to_date = me.to_date.$input.val();

			if((from_date && to_date) && to_date < from_date) {
				msgprint("To Date Must be greater than From Date")
				return false
			}
			else {
				me.single_emp_sales_details();
				me.get_mode_of_pay_details();
			}
		})
	},

	mode_of_payment_change: function() {
		var me = this;
		me.mode.$input.on("change", function() {
			me.get_mode_of_pay_details();
		})
	},



})